### PR TITLE
[Container] Unify access to FQDN

### PR DESCRIPTION
**Pull request by tiran**
FreeIPA's Python and C code used different approaches to get the FQDN of
the host. Some places assumed that gethostname() returns a FQDN. Other
code paths used glibc's resolver to resolve the current node name to a
FQDN.

Python code now uses the ipalib.constants.FQDN where a fully qualified
domain name is expected. The variable is initialized only once and avoids
potential DNS lookups.

C code uses a new helper function ipa_gethostfqdn() in util package. The
function implements similar logic as gethostfqdn() except it uses more
modern getaddrinfo(). The result is cached as well.

Fixes: https://pagure.io/freeipa/issue/8501
Signed-off-by: Christian Heimes <cheimes@redhat.com>

### CODE REVIEW

**Comment by rcritten** [comment]
```client/ipa-join.c``` and ```daemons/ipa-kdb/ipa_kdb.c``` call ```uname(2)``` to get the hostname.

In the case of ipa-join the man page will need to be updated as well since it is explicitly notes that uname() is used.

**Comment by tiran** [comment]
> `client/ipa-join.c` and `daemons/ipa-kdb/ipa_kdb.c` call `uname(2)` to get the hostname.
> 
> In the case of ipa-join the man page will need to be updated as well since it is explicitly notes that uname() is used.

Thanks! I missed the fact that some code uses uname()'s nodename. I only grepped for ``gethostname()`` and ``getfqdn``. I have updated the PR and simplified ``ipa-join``.

File: ipapython/fqdn.py
Comment (frasertweedale) [inline ipapython/fqdn.py:14]: Hmm... is this a safe assumption?

**Review by frasertweedale — COMMENTED** [review]
Initial comments, but I haven't looked at the whole diff yet (let alone tested it).  So there will be more later.

File: ipapython/fqdn.py
Comment (tiran) [reply ipapython/fqdn.py:14]: No, it's actually not safe. hostname may be partly qualified. In that case glibc/NSS can return a FQDN canonical name.

File: daemons/ipa-sam/ipa_sam.c
Comment (tiran) [reply daemons/ipa-sam/ipa_sam.c:4472]: Yes, good catch!

File: daemons/ipa-sam/ipa_sam.c
Comment (tiran) [reply daemons/ipa-sam/ipa_sam.c:4741]: Yes, good catch!

**Review by frasertweedale — COMMENTED** [review]
Finished eyeballing the C code; looking at Python code now.

File: ipaserver/install/installutils.py
Comment (frasertweedale) [inline ipaserver/install/installutils.py:126]: :+1:

File: ipapython/fqdn.py
Comment (frasertweedale) [inline ipapython/fqdn.py:29]: We should check `len(gai) > 0` and either fail with a more descriptive exception than `IndexError`, or return `hostname` (even though it may not be fully qualified).  If nssswitch is misconfigured, we should at least try and be helpful instead of failing with a cryptic error :)

**Review by frasertweedale — DISMISSED** [review]
Finished reviewing the patch; haven't tested yet.

File: ipapython/fqdn.py
Comment (tiran) [reply ipapython/fqdn.py:29]: I don't see how ``gai`` can be empty at this point. ``getaddrinfo()`` either returns at least one result or it fails with an error. All C code I have read recently always assumes that the result struct contains valid data on success.

File: util/ipa_hostname.c
Comment (tiran) [reply util/ipa_hostname.c:109]: It's a bit more radical than my initial proposal but I like it :+1:. The library is not used by any code outside FreeIPA so we are good to modify the code in the future w/o ABI issues.

File: ipapython/fqdn.py
Comment (frasertweedale) [reply ipapython/fqdn.py:29]: OK.  I prefer to be defensive ("if the return type admits it, I must handle it"), but I will leave it up to you.

**Comment by frasertweedale** [comment]
I have tested server install (not in container) and it works.  Not much testing, but something.  I will not have time to do more thorough testing before PTO.  We should test regression test trusts and, ipa-otpd->RADIUS before merging (if not adequately covered by PR-CI).  We should also test these features in a shortname environment.  I'm primarily concerned about avoiding regressions, but unless we confirm it achieves our objectives we should hold off merging it.

File: util/ipa_hostname.c
Comment (tiran) [reply util/ipa_hostname.c]: I have changed the function signature to ``const char* ipa_gethostfqdn(void)``.

File: util/ipa_hostname.c
Comment (tiran) [reply util/ipa_hostname.c]: I also wonder why this has worked ... anyhow it's now fixed. Thanks!

File: daemons/ipa-kdb/ipa_kdb.c
Comment (tiran) [reply daemons/ipa-kdb/ipa_kdb.c]: Function signature is now ``const char* ipa_gethostfqdn(void)`` and ``ipadb_context_free`` no longer frees the hostname.

File: ipapython/fqdn.py
Comment (tiran) [reply ipapython/fqdn.py:33]: I have added a comment that explains why this line cannot fail.

**Comment by frasertweedale** [comment]
@tiran 

The tests that were previously failing in #5201 now seem to be happy with this following patch.
Could you please add it to your PR (you can fix-up the tip commit "Easier to use ipa_gethostfqdn()", or keep it as a separate patch, whatever you prefer).

```patch
commit 031f62aa7d04d46f15d3b7aafb804e63632d1e93
Author: Fraser Tweedale <ftweedal@redhat.com>
Date:   Fri Oct 23 01:06:15 2020 +1100

    ipa_sam: do not clobber static buffer holding fqdn

diff --git a/daemons/ipa-sam/ipa_sam.c b/daemons/ipa-sam/ipa_sam.c
index 6993152e7..0ce858a6f 100644
--- a/daemons/ipa-sam/ipa_sam.c
+++ b/daemons/ipa-sam/ipa_sam.c
@@ -4441,7 +4441,7 @@ static char *sec_key(TALLOC_CTX *mem_ctx, const char *d)
 
 static NTSTATUS save_sid_to_secret(struct ipasam_private *ipasam_state)
 {
-       const char *hostname;
+       char hostname[IPA_HOST_FQDN_LEN];
        char *p;
        TALLOC_CTX *tmp_ctx;
        NTSTATUS status = NT_STATUS_UNSUCCESSFUL;
@@ -4466,12 +4466,13 @@ static NTSTATUS save_sid_to_secret(struct ipasam_private *ipasam_state)
                goto done;
        }
 
-       hostname = ipa_gethostfqdn();
-       if (hostname == NULL) {
+       const char *fqdn = ipa_gethostfqdn();
+       if (fqdn == NULL) {
                DEBUG(1, ("ipa_gethostfqdn failed.\n"));
                status = NT_STATUS_UNSUCCESSFUL;
                goto done;
        }
+       strncpy(hostname, fqdn, sizeof(hostname));
        p = strchr(hostname, '.');
        if (p != NULL) {
                *p = '\0';
```

**Comment by frasertweedale** [comment]
master:

* e28ec76898a0240e416910b9cfea24e82e7d91ba Unify access to FQDN
* 5d4ed65b83f6ed8bb7f3da05561698cbea692bb3 Replace nodename with ipa_gethostfqdn()
* b66b961fdd3d0b7f3c37bd245decf700e720708c Remove problematic optimization from gethostfqdn()
* 3d796a7e5102eded524e4847f8f9dd26f508ae94 Update debug strings to reflect new calls
* 727a2ffb9336a050c26c790989edaf41864c6911 Easier to use ipa_gethostfqdn()
* 3f59118ffcce07cbd74360f9ce8047e0b35960be (HEAD) ipa_sam: do not modify static buffer holding fqdn

**Pull request closed without merging**
